public/app/core/logs_model.ts

```diff
@@ -326,3 +325,141 @@ export function makeSeriesForLogs(rows: LogRowModel[], intervalMs: number): Time
+      continue;
+    }
+
+    if (getFirstTimeField(series) === -1) {
+      continue;
+    }
+
+    if (!hasFieldNamed(series, 'message')) {
+      continue;
+    }
+
+    logSeries.push(series);
+  }
+
+  const logsModel = logSeriesToLogsModel(logSeries);
+  if (logsModel) {
+    if (metricSeries.length === 0) {
+      logsModel.series = makeSeriesForLogs(logsModel.rows, intervalMs);
+    } else {
+      logsModel.series = [];
+      for (const series of metricSeries) {
+        logsModel.series.push(toLegacyResponseData(series) as TimeSeries);
+      }
+    }
+
+    return logsModel;
+  }
+
+  return undefined;
+}
+
+export function logSeriesToLogsModel(logSeries: SeriesData[]): LogsModel {
+  if (logSeries.length === 0) {
+    return undefined;
+  }
+
+  // Unique model identifier
+  const id = logSeries
+    .map(series => {
+      if (!series.labels) {
+        return '';
+      }
+
+      return Object.keys(series.labels)
+        .map(key => `key=${series.labels[key]}`)
+        .join();
+    })
+    .join();
+
+  // Find unique labels for each series
+  const commonLabels = findCommonLabels(logSeries.map(series => series.labels));
+  logSeries = logSeries.map(series => ({
+    ...series,
+    uniqueLabels: findUniqueLabels(series.labels, commonLabels),
+  }));
+
+  const uniqueLabelsPerSeries: Labels[] = [];
+  for (let n = 0; n < logSeries.length; n++) {
+    const series = logSeries[n];
+    uniqueLabelsPerSeries.push(findUniqueLabels(series.labels, commonLabels));
+  }
+
+  // Merge series entries into single list of log rows
+  const sortedRows: LogRowModel[] = _.chain(logSeries)
+    .reduce(
+      (acc: LogRowModel[], series: SeriesData, index: number) => [
+        ...acc,
+        ...series.rows.map(row => processLogSeriesRow(series, row, uniqueLabelsPerSeries[index])),
+      ],
+      []
+    )
+    .sortBy('timestamp')
+    .reverse()
+    .value();
+
+  const hasUniqueLabels = sortedRows && sortedRows.some(row => Object.keys(row.uniqueLabels).length > 0);
+
+  // Meta data to display in status
+  const meta: LogsMetaItem[] = [];
+  if (_.size(commonLabels) > 0) {
+    meta.push({
+      label: 'Common labels',
+      value: commonLabels,
+      kind: LogsMetaKind.LabelsMap,
+    });
+  }
+
+  const limits = logSeries.filter(series => series.meta && series.meta.limit);
+
+  if (limits.length > 0) {
+    meta.push({
+      label: 'Limit',
+      value: `${limits[0].meta.limit} (${sortedRows.length} returned)`,
+      kind: LogsMetaKind.String,
+    });
+  }
+
+  return {
+    id,
+    hasUniqueLabels,
+    meta,
+    rows: sortedRows,
+  };
+}
+
+export function processLogSeriesRow(series: SeriesData, row: any[], uniqueLabels: Labels): LogRowModel {
+  const ts = row[0];
+  const message = row[1];
+  const time = moment(ts);
+  const timeEpochMs = time.valueOf();
+  const timeFromNow = time.fromNow();
+  const timeLocal = time.format('YYYY-MM-DD HH:mm:ss');
+  const logLevel = getLogLevel(message);
```

Comment:
we don't need to solve this now, but it would be great to have the log level possibly come from the query:
`SELECT time,level,messsage FROM ...`

Reply:
Have added some logic for this, but probably needs some more love to support other field names than "loglevel".